### PR TITLE
HDFS-15856: Make recover the pipeline in same packet exceed times for…

### PR DESCRIPTION
**Pull request by zhuqi-lucas**
… stream closed configurable.

@jojochuang Could you help review this?
Thanks.

### CODE REVIEW

File: hadoop-hdfs-project/hadoop-hdfs/src/main/resources/hdfs-default.xml [hunk 1]
```diff
@@ -4352,6 +4352,17 @@
   </description>
 </property>
 
+<property>
+  <name>dfs.client.packet.recovery.max.times</name>
```
Comment (Hexiaoqiao) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/resources/hdfs-default.xml]: dfs.client.pipeline.recovery.max-retries?

**Review by Hexiaoqiao — COMMENTED** [review]
Thanks @qizhu-lucas for your works. It makes sense to me. Just suggest to rename variable to more readable. 
samePacketMaxRecoveryTimes -> maxPipelineRecoveryTimes?
dfs.client.packet.recovery.max.times -> dfs.client.pipeline.recovery.max-retries?

File: hadoop-hdfs-project/hadoop-hdfs/src/main/resources/hdfs-default.xml (hunk 1 above)
Comment (zhuqi-lucas) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/resources/hdfs-default.xml]: Thanks @Hexiaoqiao for review, i have fixed it in latest pull request.

File: hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java [hunk 2]
```diff
@@ -529,6 +529,7 @@ boolean doWaitForRestart() {
   private static final int CONGESTION_BACK_OFF_MAX_TIME_IN_MS =
       CONGESTION_BACKOFF_MEAN_TIME_IN_MS * 10;
   private int lastCongestionBackoffTime;
+  private int samePacketMaxRecoveryTimes;
```
Comment (Hexiaoqiao) [inline hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java]: It is better to update this variable and related method name to `maxPipelineRecoveryRetries`.

Comment (zhuqi-lucas) [reply hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java]: Thanks @Hexiaoqiao for patient review, i have fixed it in latest pull request.

File: hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java [hunk 3]
```diff
@@ -1263,14 +1265,18 @@ private boolean processDatanodeOrExternalError() throws IOException {
       packetSendTime.clear();
     }
 
-    // If we had to recover the pipeline five times in a row for the
+    // If we had to recover the pipeline exceed times which
+    // defined in maxPipelineRecoveryRetries in a row for the
```
Comment (ayushtkn) [inline hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java]: nit:
Looks some grammatical error, can we change to,
``
If we had to recover the pipeline more than the value
 defined by maxPipelineRecoveryRetries in a row for the
``

File: hadoop-hdfs-project/hadoop-hdfs/src/main/resources/hdfs-default.xml [hunk 4]
```diff
@@ -4352,6 +4352,17 @@
...
 </property>
 
+<property>
+  <name>dfs.client.pipeline.recovery.max-retries</name>
+  <value>5</value>
+  <description>
+    If we had to recover the pipeline exceed times which
+    this value defined in a row for the same packet,
+    this client likely has corrupt data or corrupting
+    during transmission.
```
Comment (ayushtkn) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/resources/hdfs-default.xml]: Same as above.

**Review by ayushtkn — COMMENTED** [review]
Thanx @qizhu-lucas for the work here. Apart from the minor nit, changes LGTM

File: hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java (hunk 3 above)
Comment (zhuqi-lucas) [reply hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java]: Thanks @ayushtkn  for review, i have fixed the grammatical error.

**Review by Hexiaoqiao — APPROVED** [review]
+1 from my side. The failed unit tests seems not related to this changes. Will commit to trunk shortly if no other comments.

**Review by ayushtkn — APPROVED** [review]

File: hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java [hunk 5]
```diff
@@ -1263,14 +1265,18 @@ private boolean processDatanodeOrExternalError() throws IOException {
...
-    if (!errorState.isRestartingNode() && ++pipelineRecoveryCount > 5) {
+    if (!errorState.isRestartingNode() && ++pipelineRecoveryCount >
+        maxPipelineRecoveryRetries) {
       LOG.warn("Error recovering pipeline for writing " +
-          block + ". Already retried 5 times for the same packet.");
+          block + ". Already retried " + maxPipelineRecoveryRetries
+          + " times for the same packet.");
       lastException.set(new IOException("Failing write. Tried pipeline " +
-          "recovery 5 times without success."));
+          "recovery "+ maxPipelineRecoveryRetries
```
Comment (jojochuang) [inline hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java]: nit: add space between " and +

File: hadoop-hdfs-project/hadoop-hdfs/src/main/resources/hdfs-default.xml [hunk 6]
```diff
@@ -4352,6 +4352,17 @@
   </description>
 </property>
 
+<property>
+  <name>dfs.client.pipeline.recovery.max-retries</name>
+  <value>5</value>
+  <description>
+    If we had to recover the pipeline more than the value
```
Comment (jojochuang) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/resources/hdfs-default.xml]: Suggest to add a description for this configuration property. Something like "if the DFS client encounters errors in write pipeline, retry up to the number defined by this property before giving up"

File: hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java (hunk 5 above)
Comment (zhuqi-lucas) [reply hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java]: Thanks @jojochuang for review.
Fixed it latest pull request.

Comment (qizhu-lucas) [reply hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java]: Thanks @jojochuang for review.
Fixed it latest pull request.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/resources/hdfs-default.xml (hunk 6 above)
Comment (zhuqi-lucas) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/resources/hdfs-default.xml]: Good suggestion, i have changed to it.
Thanks @jojochuang

Comment (qizhu-lucas) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/resources/hdfs-default.xml]: Good suggestion, i have changed to it.
Thanks @jojochuang

File: hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java (hunk 5 above)
Comment (qizhu-lucas) [reply hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java]: Thanks @jojochuang for review.
Fixed it latest pull request.

**Comment by zhuqi-lucas** [comment]
I think the test failed is not related to this.

**Comment by zhuqi-lucas** [comment]
@jojochuang 
Could you help check the latest patch? 
Thanks.

**Comment by Hexiaoqiao** [comment]
Committed to trunk. Thanks @qizhu-lucas for your contributions. Thanks @ayushtkn and @jojochuang for your reviews.

**Pull request merged**